Document HTTP service (#42331)
elastic/kibana: src/core/MIGRATION.md

  }
```

### Handle HTTP request with New Platform HTTP Service
Kibana HTTP Service provides own abstraction for work with HTTP stack.
Plugins don't have direct access to `hapi` server and its primitives anymore. Moreover,

> you shouldn't rely on the fact that HTTP Service uses one or another library under the hood, there are high chances that we can switch to another stack at all. If HTTP Service

Reviewer: ```suggestion
plugins shouldn't rely on the fact that HTTP Service uses one or another library under the hood. This gives the platform flexibility to upgrade or changing our internal HTTP stack without breaking plugins. If the HTTP Service
```

> lacks needed functionality we happy to discuss and support your needs.

Reviewer: ```suggestion
lacks functionality you need, we are happy to discuss and support your needs.
```

To handle an incoming request in your plugin you should:

> - create Router instance. Use `plugin id` as a prefix path segment for your routes.

Reviewer: ```suggestion
- create a `Router` instance. Use `plugin-id` as a prefix path segment for your routes.
```

**optional:** we could provide a getRouter function to plugins that automatically uses their plugin id as the prefix?

Reviewer 2: yes, we can. `Router` is a class exported from `src/core/server`. because `Router` belongs to the `HTTP` domain, we can make it a part of `HTTP service` contract and expose a factory function which already prefixed with `plugin-id`.

```ts
import { Router } from 'src/core/server';
const router = new Router('my-app');
```

> - use `@kbn/config-schema` package to create a schema to validate `request params`, `query`, `body` if an incoming request used them to pass additional details. Every incoming request will be validated against the created schema. If validation failed, the request is rejected with `400` status and `Bad request` error.

Reviewer: ```suggestion
- use `@kbn/config-schema` package to create a schema to validate the request `params`, `query`, and `body`. Every incoming request will be validated against the created schema. If validation failed, the request is rejected with `400` status and `Bad request` error without calling the route's handler.
```

To opt out of validating the request, specify `false`.
```ts
import { schema, TypeOf } from '@kbn/config-schema';
const validate = {
  params: schema.object({
    id: schema.string(),
  }),
};
```

- declare a function to respond to incoming request.
The function will receive `request` object containing request details: url, headers, matched route, as well as validated `params`, `query`, `body`.
And `response` object instructing HTTP server to create HTTP response with information sent back to the client as the response body, headers, and HTTP status.

> Unlike, `hapi` route handler in the Legacy platform, any exception raised during the handler call will generate `500 Server error` response and log error details for further investigation.

Reviewer: ```suggestion
Unlike, `hapi` route handler in the Legacy platform, any exception raised during the handler call will generate `500 Server error` response and log error details for further investigation. See below for returning custom error responses.
```

```ts
  const handler = async (request: KibanaRequest, response: ResponseFactory) => {
    const data = await findObject(request.params.id);
    // creates a command to respond with 'not found' error
    if (!data) return response.notFound();
    // creates a command to send found data to the client and set response headers
    return response.ok(data, {
      headers: {
        'content-type': 'application/json'
      }
    });
  }
```

- register route handler for GET request to 'my-app/path/{id}' path
```ts
import { schema, TypeOf } from '@kbn/config-schema';
import { Router } from 'src/core/server';
const router = new Router('my-app');

const validate = {
  params: schema.object({
    id: schema.string(),
  }),
};

router.get(
  {
    path: 'path/{id}',
    validate
  },
  async (request, response) => {
    const data = await findObject(request.params.id);
    if (!data) return response.notFound();
    return response.ok(data, {
      headers: {
        'content-type': 'application/json'
      }
    });
  }
);
```
#### What data I can send back with a response?
You have several options to create a response utilizing `response`:

> 1. create a successful response. Supported types of response body are:

Reviewer: Can we capitalize these list items?

  headers: {
    'content-type': 'application/json'
  }
});
```
2. create redirect response. Redirection URL is configures via 'Location' header.
```js
return response.redirected('The document has moved', {
  headers: {
    location: '/new-url',
  },
});
```
3. create error response. You may pass an error message to the client, where error message can be:
- `string` send message text
- `Error` send the message text of given Error object.
- `{ message: string | Error, meta: {data: Record<string, any>, ...} }` - send message text and attach additional error metadata.
```js
return response.unauthorized('User has no access to the requested resource.', {
  headers: {
    'WWW-Authenticate': 'challenge',
  }
})
return response.badRequest();
return response.badRequest('validation error');

try {
  // ...
} catch(error){
 return response.badRequest(error);
}

return response.badRequest({
  message: 'validation error',
  meta: {
    data: {
      requestBody: request.body,
      failedFields: validationResult
    },
  }
});

try {
  // ...
} catch(error){
 return response.badRequest({
  message: error,
  meta: {
    data: {
      requestBody: request.body,
    },
  }
});
}

```

> 4. create a custom response. It might happen that `response factory` doesn't cover your use case and you want to specify HTTP response status code as well.

Reviewer: ```suggestion
4. create a custom response. `ResponseFactory` may not cover your use case, so you can use the `custom` function to customize the response.
```

```js
return response.custom('ok', {
  statusCode: 201,
  headers: {
    location: '/created-url'
  }
})

> ```

Reviewer: What about testing? Could you provide an example of how to test these handler methods?

Reviewer 2: Testing is not relevant as I moved the docs closer to the code. 
We can provide a separate doc recommending test patterns for NP.
Although I don't it's common for plugins to test their endpoints in insolation. Usually, It's a part of integration/functional tests.

Reviewer: > Although I don't it's common for plugins to test their endpoints in insolation.

Heh this really made upgrading Hapi very hard last time. IMO, endpoints should be tested with decent units so there can be fewer integrations tests needed. Anyways, I'm fine with not doing that here in this PR.
